README.md
# dockerized-chef-services
Docker definitions for Chef Server and Automate 1.x

![Architecture Diagram](https://www.lucidchart.com/publicSegments/view/4f01dc86-c34a-49a9-b619-f3d1056e7a41/image.png)

## Release Notes
Please read the release notes [here](https://github.com/chef-customers/dockerized-chef-services/blob/master/RELEASE_NOTES.md) for any notable changes.

## How it works

* All of the Chef Server and Chef Automate services have been packaged as Habitat `.hart` files
* The hab packages have been exported to docker containers and published to docker hub
* These containers support running under a random, arbitrary effective user and/or group id
* Launch containers via the included [docker-chef.sh](https://github.com/chef-customers/dockerized-chef-services/tree/master/terraform/docker-chef.sh) script.
* Each host forms its own non-permanent Habitat gossip ring, sharing service discovery data intra-host only
* Environment variables are used to configure settings from the default

## Deploying it

### Using AWS and Terraform

* Customize the `terraform.tfvars.example` [Terraform](https://terraform.io) configuration file, rename it to `terraform.tfvars` and deploy with `terraform apply`
* Sit back and enjoy - in about 4 minutes you'll have a cluster up and running

```bash
git clone https://github.com/chef-customers/dockerized-chef-services.git
cd dockerized-chef-services/terraform
cp terraform.tfvars.example terraform.tfvars
vi terraform.tfvars # change anything that needs customization
terraform init
terraform apply
```

### Manual

* Provision two hosts as depicted in the Architecture diagram above and ensure the items from [Deployment notes](https://github.com/chef-customers/dockerized-chef-services#deployment-notes)
* Copy `docker-run.sh` to each host
* Create a customized `env.sh` and place in the same directory as `docker-run.sh`
* Create the `$DATA_MOUNT` mount point on each host and ensure that it's owned by `$USER_ID:$GROUP_ID`
* Copy a `passwd` and `group` file (see examples below) into `$DATA_MOUNT` and ensure they're readable by `$USER_ID:$GROUP_ID`
* Run `docker-run.sh -h` to see the help menu

```
$ ./docker-chef.sh -h
This is a control script for starting|stopping Chef Server and Chef Automate docker services.

You must specify the following options:
 -s [automate|chef-server]           REQUIRED: Services type: Chef Server or Chef Automate
 -a [stop|start]                     REQUIRED: Action type: start or stop services
 -n [container name]                 OPTIONAL: The docker container name. Leaving blank implies ALL
 -h                                  OPTIONAL: Print this help message

 ex. ./docker-chef.sh -s chef-server -a start            # starts up all Chef Server services
 ex. ./docker-chef.sh -s automate -a stop -n logstash    # stops Automate's logstash service

$
```

Example `group`

```
root:x:0:
bin:x:1:daemon
sys:x:2:
kmem:x:3:
tape:x:4:
tty:x:5:
daemon:x:6:
floppy:x:7:
disk:x:8:
lp:x:9:
dialout:x:10:
audio:x:11:
video:x:12:
utmp:x:13:
usb:x:14:
cdrom:x:15:
adm:x:16:
messagebus:x:18:
systemd-journal:x:23:
input:x:24:
mail:x:34:
nogroup:x:99:
users:x:999:
hab:x:42:hab:
chef-dev-ux:x:8888:
```

Example `passwd`

```
root:x:0:0:root:/root:/bin/sh
bin:x:1:1:bin:/dev/null:/bin/false
daemon:x:6:6:Daemon User:/dev/null:/bin/false
nobody:x:99:99:Unprivileged User:/dev/null:/bin/false
hab:x:42:42:hab User:/:/bin/false
chef-dev-ux:x:9999:8888:Test User:/:/bin/false
```

Example `env.sh`:

```bash
#!/bin/bash

export LC_ALL=en_US.UTF-8

export USER_ID=9999
export GROUP_ID=8888
export DATA_MOUNT=/mnt/data
export ENTERPRISE=brewinc
export ADMIN_PASSWORD=insecurepassword
export HOST_IP=$(hostname --ip-address)
export AUTOMATE_ENABLED=true
export AUTOMATE_SERVER=172.31.20.110
export AUTOMATE_TOKEN=93a49a4f2482c64126f7b6015e6b0f30284287ee4054ff8807fb63d9cbd1c506
export CHEF_SERVER_DOCKER_ORIGIN=chefserverofficial
export CHEF_SERVER_VERSION=stable
export AUTOMATE_DOCKER_ORIGIN=chefdemo
export AUTOMATE_VERSION=stable
export DOCKER_REQUIRES_SUDO=false
export DOCKER_DETACH_CONTAINER=true
```

### Docker-compose

```bash
git clone https://github.com/chef-customers/dockerized-chef-services.git
cd dockerized-chef-services/docker
docker-compose -f chef-server.yml down && docker system prune --volumes -f && docker-compose -f chef-server.yml up
docker-compose -f automate.yml down && docker system prune --volumes -f && docker-compose -f automate.yml up
```

## Deployment notes

Some environments may require the following, particularly if Elasticsearch refuses to start:

```
sudo cat > /etc/sysctl.d/00-chef.conf <<EOF
vm.swappiness=10
vm.max_map_count=262144
vm.dirty_ratio=20
vm.dirty_background_ratio=30
vm.dirty_expire_centisecs=30000
EOF
sudo sysctl -p /etc/sysctl.d/00-chef.conf

sudo cat > /etc/security/limits.d/20-nproc.conf<<EOF
*   soft  nproc     65536
*   hard  nproc     65536
*   soft  nofile    1048576
*   hard  nofile    1048576
EOF
```

Ensuring that locale is UTF-8 is necessary if not already set:

```
export LC_ALL=en_US.UTF-8
```

For the non-root images, all the services run on non-privileged ports. To enable seamless external access
to TCP ports 80 and 443 the following iptables rules must be applied.
Alternatively, use a Load Balancer.

If not using a LB, on the Chef Server Host run:

```
sudo iptables -A PREROUTING -t nat -p tcp --dport 80 -j REDIRECT --to-port 8080
sudo iptables -A PREROUTING -t nat -p tcp --dport 443 -j REDIRECT --to-port 8443
sudo iptables -t nat -A OUTPUT -o lo -p tcp --dport 80 -j REDIRECT --to-port 8080
sudo iptables -t nat -A OUTPUT -o lo -p tcp --dport 443 -j REDIRECT --to-port 8443
```

And on the Automate Host:

```
sudo iptables -A PREROUTING -t nat -p tcp --dport 80 -j REDIRECT --to-port 8080
sudo iptables -A PREROUTING -t nat -p tcp --dport 443 -j REDIRECT --to-port 8443
```

# Operating it

## Starting | Stopping

For this use the `docker-chef.sh` script as shown above.

## Administrative Functions

### Chef Server

Run the functional test suite to ensure Chef server is working:

```
docker exec -it chef-server-ctl chef-server-test
```

Add users, orgs, etc to the Chef server

docker run:

```
docker exec -it chef-server-ctl chef-server-ctl (subcommands)
```

### Chef Automate

Coming Soon! We will be releasing an `automat-ctl` command in the `workflow-server` container.

## Logging
All container logs are directed to STDOUT. You should employ a Docker logging mechanism to ensure those logs are captured and aggregated in a central location. Being able to provide full logs from all containers is necessary in order to recieve support.

To view the status of any container:

docker run:

```
docker logs <container name or id>
```

An easy way to check on the health of Chef Server and Automate is to look for the health checks from `oc_erchef` and `workflow-server` containers accordingly. They will look like:

```
oc_erchef.default hook[health_check]:(HK): {"status":"pong","upstreams":{"chef_elasticsearch":"pong","chef_sql":"pong","oc_chef_authz":"pong"},
"keygen":{"keys":10,"max":10,"max_workers":10,"cur_max_workers":10,"inflight":0,"avail_workers":10,"start_size":0},"indexing":{"mode":"batch"}}
```

```
workflow-server.default hook[health_check]:(HK): {"status":"pong","configuration_mode":"standalone","fips_mode":"false","upstreams":
[{"postgres":{"status":"pong"},"lsyncd":{"status":"not_running"},"elasticsearch":{"status":"pong","cluster_name":"elasticsearch","status":
"yellow","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":20,"active_shards":20,"relocating_shards":0,
"initializing_shards":0,"unassigned_shards":20,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,
"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":50.0},"rabbitmq":{"status":"pong","vhost_aliveness":{"status":
"pong"},"node_health":{"status":"pong"}}}]}
```


RELEASE_NOTES.md
This document contains historical release notes from the last minor
release and prior.  The current release notes and subsequent patch
releases can be found [RELEASE\_NOTES.md](RELEASE_NOTES.md).

## 0.3.0 (2018-04-12)
  * Minor correction to `stop_all` function in order to prevent stopping non-related services


## 0.2.0 (2018-04-12)
  * Tested and validated with Chef Server 12.17.41 and Automate 1.8.38

  **NOTE**: Set `$CHEF_SERVER_VERSION` and `$AUTOMATE_VERSION` repsectively to pull these verified versions from Dockerhub

  * Introduced a single control script `docker-chef.sh` so that start|stop and `/hab/sup` directory clearing functionality is consistent.

  **NOTE**: Please use this script as it is how we can better enforce desired behavior. For example, you no longer have to create all the `/hab/sup` and `/hab/svc` directories, just the `$DATA_MOUNT` - the script does the rest.

  * `automate-ctl` functionality is working and will be made available in the next release.


terraform/chef-server.sh
#!/bin/bash

# Configurable shell environment variables:
# CHEF_SERVER_DOCKER_ORIGIN - denotes the docker origin (dockerhub ID) or default to `chefserverofficial`
# AUTOMATE_DOCKER_ORIGIN - denotes the docker origin (dockerhub ID) or default to `chefdemo`
# CHEF_SERVER_VERSION -  the version identifier tag on the Chef Server packages
# AUTOMATE_VERSION -  the version identifier tag on the postgresql and elasticsearch packages from the `chefdemo` docker origin
# HOST_IP - the IP address of the docker host. 172.17.0.1 is commonly the docker0 interface which is fine
# AUTOMATE_ENABLED - enable the Automate data collector (true or false)
# AUTOMATE_SERVER - the IP address or hostname of the Automate server
# AUTOMATE_TOKEN - the token for the Automate server data collector
# DATA_MOUNT - the mount point for the data
# USER_ID - the user ID to use (numeric)
# GROUP_ID - the group ID to use (numeric)
#
# The above variables should all be set in a file named env.sh that lives beside this script.

THISDIR="$(dirname "$(which "$0")")"
if [ -f "${THISDIR}/env.sh" ]; then
 echo "Setting ENVIRONMENT variables"
 . $THISDIR/env.sh
fi

docker_svc_start () {
  echo "Starting $1"
  dirs="${DATA_MOUNT:-/mnt/hab}/${1}_svc ${DATA_MOUNT:-/mnt/hab}/${1}_sup"
  echo "Ensuring $dirs directories exist and removing stale LOCK files"
  mkdir -p $dirs
  rm -f ${DATA_MOUNT:-/mnt/hab}/${1}_sup/default/LOCK
  docker run --rm -it \
    --name="${1}" \
    --env="${4:-ILOVECHEF=1}" \
    --volume ${DATA_MOUNT:-/mnt/hab}/passwd:/etc/passwd:ro \
    --volume ${DATA_MOUNT:-/mnt/hab}/group:/etc/group:ro \
    --volume ${DATA_MOUNT:-/mnt/hab}/${1}_svc:/hab/svc \
    --volume ${DATA_MOUNT:-/mnt/hab}/${1}_sup:/hab/sup \
    --cap-drop="NET_BIND_SERVICE" \
    --cap-drop="SETUID" \
    --cap-drop="SETGID" \
    --ulimit nofile=65536:65536 \
    --user="${USER_ID:-42}:${GROUP_ID:-42}" \
    --network=host \
    --detach=true \
    $2 \
    $3
}

declare -A postgresql
postgresql["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/postgresql:${AUTOMATE_VERSION:-stable}"
postgresql["env"]="HAB_POSTGRESQL=[superuser]
name = 'hab'
password = 'chefrocks'
"
postgresql["supargs"]=""

declare -A chef_server_ctl
chef_server_ctl["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/chef-server-ctl:${CHEF_SERVER_VERSION:-stable}"
chef_server_ctl["env"]="HAB_CHEF_SERVER_CTL=[chef_server_api]
ip = \"${HOST_IP:-172.17.0.1}\"
ssl_port = "8443"
[secrets.data_collector]
token = \"${AUTOMATE_TOKEN:-93a49a4f2482c64126f7b6015e6b0f30284287ee4054ff8807fb63d9cbd1c506}\"
"
chef_server_ctl["supargs"]="--peer ${HOST_IP:-172.17.0.1} --listen-gossip 0.0.0.0:9650 --listen-http 0.0.0.0:9660"

declare -A elasticsearch
elasticsearch["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/elasticsearch5:${AUTOMATE_VERSION:-stable}"
elasticsearch["env"]=""
elasticsearch["supargs"]="--peer ${HOST_IP:-172.17.0.1} --listen-gossip 0.0.0.0:9651 --listen-http 0.0.0.0:9661"

declare -A oc_id
oc_id["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_id:${CHEF_SERVER_VERSION:-stable}"
oc_id["env"]=""
oc_id["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9652 --listen-http 0.0.0.0:9662"

declare -A bookshelf
bookshelf["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/bookshelf:${CHEF_SERVER_VERSION:-stable}"
bookshelf["env"]=""
bookshelf["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9653 --listen-http 0.0.0.0:9663"

declare -A oc_bifrost
oc_bifrost["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_bifrost:${CHEF_SERVER_VERSION:-stable}"
oc_bifrost["env"]=""
oc_bifrost["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9654 --listen-http 0.0.0.0:9664"

declare -A oc_erchef
oc_erchef["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_erchef:${CHEF_SERVER_VERSION:-stable}"
oc_erchef["env"]="HAB_OC_ERCHEF=[data_collector]
enabled = ${AUTOMATE_ENABLED:-false}
server = \"${AUTOMATE_SERVER:-localhost}\"
port = 443
[chef_authn]
keygen_cache_workers = 2
keygen_cache_size = 10
keygen_start_size = 0
keygen_timeout = 20000
"
oc_erchef["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind bookshelf:bookshelf.default --bind oc_bifrost:oc_bifrost.default --bind database:postgresql.default --bind elasticsearch:elasticsearch5.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9655 --listen-http 0.0.0.0:9665"

declare -A chef_server_nginx
chef_server_nginx["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/chef-server-nginx:${CHEF_SERVER_VERSION:-stable}"
chef_server_nginx["env"]=""
chef_server_nginx["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind oc_erchef:oc_erchef.default --bind oc_bifrost:oc_bifrost.default --bind oc_id:oc_id.default --bind bookshelf:bookshelf.default --bind elasticsearch:elasticsearch5.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9656 --listen-http 0.0.0.0:9666"

stop_svc () {
  echo "Stopping $1"
  docker stop $1 >/dev/null 2>&1 || true
}

stop_all () {
  echo "Stopping ALL.."
  docker stop $(docker ps -aq) >/dev/null 2>&1 || true
  echo "Removing ${DATA_MOUNT:-/mnt/hab}/*_sup"
  rm -rf ${DATA_MOUNT:-/mnt/hab}/*_sup
}

start_all () {
  docker_svc_start "postgresql" "${postgresql[image]}" "${postgresql[supargs]}" "${postgresql[env]}"
  docker_svc_start "chef-server-ctl" "${chef_server_ctl[image]}" "${chef_server_ctl[supargs]}" "${chef_server_ctl[env]}"
  docker_svc_start "elasticsearch" "${elasticsearch[image]}" "${elasticsearch[supargs]}" "${elasticsearch[env]}"
  docker_svc_start "oc_id" "${oc_id[image]}" "${oc_id[supargs]}" "${oc_id[env]}"
  docker_svc_start "bookshelf" "${bookshelf[image]}" "${bookshelf[supargs]}" "${bookshelf[env]}"
  docker_svc_start "oc_bifrost" "${oc_bifrost[image]}" "${oc_bifrost[supargs]}" "${oc_bifrost[env]}"
  docker_svc_start "oc_erchef" "${oc_erchef[image]}" "${oc_erchef[supargs]}" "${oc_erchef[env]}"
  docker_svc_start "chef-server-nginx" "${chef_server_nginx[image]}" "${chef_server_nginx[supargs]}" "${chef_server_nginx[env]}"
}

case "$1" in
  stop)
    case "$2" in
      "")
        stop_all
        ;;
      *)
        stop_svc $2
        ;;
    esac
    ;;
  start)
    case "$2" in
      "")
        start_all
        ;;
      *)
      # https://stackoverflow.com/questions/39297530/bash-use-variable-as-name-of-associative-array-when-calling-value
      svc=$(echo $2|tr '-' '_')
      image=$svc[image]
      supargs=$svc[supargs]
      env=$svc[env]
      docker_svc_start "$2" "${!image}" "${!supargs}" "${!env}"
      ;;
    esac
  ;;
esac

terraform/docker-chef.sh
#!/bin/bash

# Configurable shell environment variables:
#
# CHEF_SERVER_DOCKER_ORIGIN - denotes the docker origin (dockerhub ID) or default to `chefserverofficial`
# AUTOMATE_DOCKER_ORIGIN - denotes the docker origin (dockerhub ID) or default to `chefdemo`
# CHEF_SERVER_VERSION -  the version identifier tag on the Chef Server packages
# AUTOMATE_VERSION -  the version identifier tag on the postgresql and elasticsearch packages from the `chefdemo` docker origin
# HOST_IP - the IP address of the docker host. 172.17.0.1 is commonly the docker0 interface which is fine
# AUTOMATE_ENABLED - enable the Automate data collector (true or false)
# AUTOMATE_SERVER - the IP address or hostname of the Automate server
# AUTOMATE_TOKEN - the token for the Automate server data collector
# ENTERPRISE - the name of the Automate enterprise to create
# ADMIN_PASSWORD - the initial password to set for the 'admin' user in the Automate UI
# DATA_MOUNT - the mount point for the data
# USER_ID - the user ID to use (numeric)
# GROUP_ID - the group ID to use (numeric)
# DOCKER_REQUIRES_SUDO - [true|false] whether or not docker requires sudo to run when invoked by the user running this script

# The above variables should all be set in a file named env.sh that lives beside this script.
THISDIR="$(dirname "$(which "$0")")"
if [ -f "${THISDIR}/env.sh" ]; then
 . "${THISDIR}/env.sh"
fi

banner="This is a control script for starting|stopping Chef Server and Chef Automate docker services.

You must specify the following options:
 -s [automate|chef-server]           REQUIRED: Services type: Chef Server or Chef Automate
 -a [stop|start]                     REQUIRED: Action type: start or stop services
 -n [container name]                 OPTIONAL: The docker container name. Leaving blank implies ALL
 -h                                  OPTIONAL: Print this help message

 ex. $0 -s chef-server -a start            # starts up all Chef Server services
 ex. $0 -s automate -a stop -n logstash    # stops Automate's logstash service

"

usage () {
  echo "$banner"
  exit 1
}

if [ $# -eq 0 ]; then
  usage
fi

while getopts "s:a:n:h" opt; do
  case $opt in
    s)
      echo "Service type: $OPTARG"
      export SERVICE_TYPE=$OPTARG
      ;;
    a)
      echo "Service action: $OPTARG"
      export SERVICE_ACTION=$OPTARG
      ;;
    n)
      echo "Service name: $OPTARG"
      export SERVICE_NAME=$OPTARG
      ;;
    h)
      usage
      ;;
    \?)
      echo "Invalid option: -$OPTARG" >&2
      exit 1
      ;;
    :)
      echo "Option -$OPTARG requires an argument." >&2
      exit 1
      ;;
  esac
done

# Service Definitinons
#

declare -A postgresql
postgresql["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/postgresql:${AUTOMATE_VERSION:-stable}"
postgresql["env"]="HAB_POSTGRESQL=[superuser]
name = 'hab'
password = 'chefrocks'
"
postgresql["supargs"]=""

declare -A elasticsearch
elasticsearch["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/elasticsearch5:${AUTOMATE_VERSION:-stable}"
elasticsearch["env"]=""
elasticsearch["supargs"]="--peer ${HOST_IP:-172.17.0.1} --listen-gossip 0.0.0.0:9651 --listen-http 0.0.0.0:9661"

# Chef Server
#
declare -A chef_server_ctl
chef_server_ctl["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/chef-server-ctl:${CHEF_SERVER_VERSION:-stable}"
chef_server_ctl["env"]="HAB_CHEF_SERVER_CTL=[chef_server_api]
ip = \"${HOST_IP:-172.17.0.1}\"
ssl_port = \"8443\"
[secrets.data_collector]
token = \"${AUTOMATE_TOKEN:-93a49a4f2482c64126f7b6015e6b0f30284287ee4054ff8807fb63d9cbd1c506}\"
"
chef_server_ctl["supargs"]="--peer ${HOST_IP:-172.17.0.1} --listen-gossip 0.0.0.0:9650 --listen-http 0.0.0.0:9660"

declare -A oc_id
oc_id["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_id:${CHEF_SERVER_VERSION:-stable}"
oc_id["env"]=""
oc_id["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9652 --listen-http 0.0.0.0:9662"

declare -A bookshelf
bookshelf["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/bookshelf:${CHEF_SERVER_VERSION:-stable}"
bookshelf["env"]=""
bookshelf["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9653 --listen-http 0.0.0.0:9663"

declare -A oc_bifrost
oc_bifrost["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_bifrost:${CHEF_SERVER_VERSION:-stable}"
oc_bifrost["env"]=""
oc_bifrost["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9654 --listen-http 0.0.0.0:9664"

declare -A oc_erchef
oc_erchef["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/oc_erchef:${CHEF_SERVER_VERSION:-stable}"
oc_erchef["env"]="HAB_OC_ERCHEF=[data_collector]
enabled = ${AUTOMATE_ENABLED:-false}
server = \"${AUTOMATE_SERVER:-localhost}\"
port = 443
[chef_authn]
keygen_cache_workers = 2
keygen_cache_size = 10
keygen_start_size = 0
keygen_timeout = 20000
"
oc_erchef["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind bookshelf:bookshelf.default --bind oc_bifrost:oc_bifrost.default --bind database:postgresql.default --bind elasticsearch:elasticsearch5.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9655 --listen-http 0.0.0.0:9665"

declare -A chef_server_nginx
chef_server_nginx["image"]="${CHEF_SERVER_DOCKER_ORIGIN:-chefserverofficial}/chef-server-nginx:${CHEF_SERVER_VERSION:-stable}"
chef_server_nginx["env"]=""
chef_server_nginx["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind oc_erchef:oc_erchef.default --bind oc_bifrost:oc_bifrost.default --bind oc_id:oc_id.default --bind bookshelf:bookshelf.default --bind elasticsearch:elasticsearch5.default --bind chef-server-ctl:chef-server-ctl.default --listen-gossip 0.0.0.0:9656 --listen-http 0.0.0.0:9666"

# Automate
#

declare -A rabbitmq
rabbitmq["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/rabbitmq:${AUTOMATE_VERSION:-stable}"
rabbitmq["env"]="HAB_RABBITMQ=[rabbitmq]
default_vhost = '/insights'
default_user = 'insights'
default_pass = 'chefrocks'
[rabbitmq.management]
enabled = true
"
rabbitmq["supargs"]="--peer ${HOST_IP:-172.17.0.1} --listen-gossip 0.0.0.0:9650 --listen-http 0.0.0.0:9660"

declare -A logstash
logstash["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/logstash:${AUTOMATE_VERSION:-stable}"
logstash["env"]=""
logstash["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind elasticsearch:elasticsearch5.default --bind rabbitmq:rabbitmq.default --listen-gossip 0.0.0.0:9652 --listen-http 0.0.0.0:9662"

declare -A workflow_server
workflow_server["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/workflow-server:${AUTOMATE_VERSION:-stable}"
workflow_server["env"]="HAB_WORKFLOW_SERVER=
enterprise = \"${ENTERPRISE:-default}\"
default_admin_password = \"${ADMIN_PASSWORD:-chefrocks}\"
[data_collector]
token = \"${AUTOMATE_TOKEN:-93a49a4f2482c64126f7b6015e6b0f30284287ee4054ff8807fb63d9cbd1c506}\"
[mlsa]
accept = true
"
workflow_server["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind database:postgresql.default --bind elasticsearch:elasticsearch5.default --bind rabbitmq:rabbitmq.default --listen-gossip 0.0.0.0:9653 --listen-http 0.0.0.0:9663"

declare -A notifications
notifications["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/notifications:${AUTOMATE_VERSION:-stable}"
notifications["env"]=""
notifications["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind elasticsearch:elasticsearch5.default --bind rabbitmq:rabbitmq.default --listen-gossip 0.0.0.0:9654 --listen-http 0.0.0.0:9664"

declare -A compliance
compliance["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/compliance:${AUTOMATE_VERSION:-stable}"
compliance["env"]="HAB_COMPLIANCE_SERVICE=[service]
host = \"0.0.0.0\"
[profiles]
secrets_key = \"12345678901234567890123456789012\"
"
compliance["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind postgresql:postgresql.default --bind elasticsearch:elasticsearch5.default --bind workflow:workflow-server.default --listen-gossip 0.0.0.0:9655 --listen-http 0.0.0.0:9665"

declare -A automate_nginx
automate_nginx["image"]="${AUTOMATE_DOCKER_ORIGIN:-chefdemo}/automate-nginx:${AUTOMATE_VERSION:-stable}"
automate_nginx["env"]="HAB_AUTOMATE_NGINX=
port = ${PILOT_HTTP_PORT:-8080}
ssl_port = ${PILOT_HTTPS_PORT:-8443}
[mlsa]
accept = true
"
automate_nginx["supargs"]="--peer ${HOST_IP:-172.17.0.1} --bind compliance:compliance.default --bind elasticsearch:elasticsearch5.default --bind workflow:workflow-server.default --bind notifications:notifications.default --listen-gossip 0.0.0.0:9656 --listen-http 0.0.0.0:9666"

# Service functions
#

sudo_cmd () {
  if [ "$DOCKER_REQUIRES_SUDO" == "true" ]; then
    echo "sudo -E "
  else
    echo ""
  fi
}

docker_svc_start () {
  # https://stackoverflow.com/questions/39297530/bash-use-variable-as-name-of-associative-array-when-calling-value
  svc=$(echo "$1"|tr '-' '_')
  image="$svc[image]"
  supargs="$svc[supargs]"
  env="$svc[env]"

  echo "Starting ${!image}"
  dirs="${DATA_MOUNT:-/mnt/hab}/${1}_svc ${DATA_MOUNT:-/mnt/hab}/${1}_sup"
  echo "Ensuring $dirs directories exist and removing stale LOCK files"
  mkdir -p $dirs
  rm -f ${DATA_MOUNT:-/mnt/hab}/${1}_sup/default/LOCK
  $(sudo_cmd) docker run --rm -it \
    --name="${1}" \
    --env="HOME=/hab/svc/${1}/data" \
    --env="${!env:-ILOVECHEF=1}" \
    --volume ${DATA_MOUNT:-/mnt/hab}/passwd:/etc/passwd:ro \
    --volume ${DATA_MOUNT:-/mnt/hab}/group:/etc/group:ro \
    --volume ${DATA_MOUNT:-/mnt/hab}/${1}_svc:/hab/svc \
    --volume ${DATA_MOUNT:-/mnt/hab}/${1}_sup:/hab/sup \
    --cap-drop="NET_BIND_SERVICE" \
    --cap-drop="SETUID" \
    --cap-drop="SETGID" \
    --ulimit nofile=65536:65536 \
    --user="${USER_ID:-42}:${GROUP_ID:-42}" \
    --network=host \
    --detach=${DOCKER_DETACH_CONTAINER:-true} \
    ${!image} \
    ${!supargs}
}

stop_svc () {
  echo "Stopping $1"
  $(sudo_cmd) docker stop "$1" >/dev/null 2>&1 || true
}

stop_all () {
  echo "Stopping ALL.."
  case "$1" in
    chef-server)
      for svc in postgresql chef-server-ctl elasticsearch oc_id bookshelf oc_bifrost oc_erchef chef-server-nginx; do
        stop_svc "$svc"
      done
      ;;
    automate)
      for svc in postgresql rabbitmq elasticsearch logstash workflow-server notifications compliance automate-nginx; do
        stop_svc "$svc"
      done
      ;;
    *)
      echo "not implemented"
      ;;
  esac
  echo "Removing ${DATA_MOUNT:-/mnt/hab}/*_sup"
  rm -rf ${DATA_MOUNT:-/mnt/hab}/*_sup
}

start_all () {
  case "$1" in
    chef-server)
      for svc in postgresql chef-server-ctl elasticsearch oc_id bookshelf oc_bifrost oc_erchef chef-server-nginx; do
        docker_svc_start "$svc"
      done
      ;;
    automate)
      for svc in postgresql rabbitmq elasticsearch logstash workflow-server notifications compliance automate-nginx; do
        docker_svc_start "$svc"
      done
      ;;
    *)
      echo "not implemented"
      ;;
  esac
}

case "$SERVICE_TYPE" in
  # chef-server or automate
  "")
     usage
     ;;
  automate|chef-server)
    # start or stop
    case "$SERVICE_ACTION" in
      stop)
        case "$SERVICE_NAME" in
          # optional service name
          "")
            stop_all "$SERVICE_TYPE"
            ;;
          *)
            stop_svc "$SERVICE_NAME"
            ;;
        esac
        ;;
      start)
        case "$SERVICE_NAME" in
          # optional service name
          "")
            start_all "$SERVICE_TYPE"
            ;;
          *)
          docker_svc_start "$SERVICE_NAME"
          ;;
        esac
        ;;
      *)
        usage
        ;;
    esac
      ;;
  *)
    usage
    ;;
esac
